### PR TITLE
HBASE-23035 Retain region to the last RegionServer make the failover …

### PR DESCRIPTION
**Pull request by infraio**
…slower

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m 20s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :yellow_heart: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   7m 53s |  master passed  |
| :green_heart: |  compile  |   1m 18s |  master passed  |
| :green_heart: |  checkstyle  |   1m 47s |  master passed  |
| :green_heart: |  shadedjars  |   6m  5s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 43s |  master passed  |
| :blue_heart: |  spotbugs  |   5m 12s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   5m 11s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   7m  5s |  the patch passed  |
| :green_heart: |  compile  |   1m 15s |  the patch passed  |
| :green_heart: |  javac  |   1m 15s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 43s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m 57s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  21m 36s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 47s |  the patch passed  |
| :green_heart: |  findbugs  |   5m 31s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 275m 35s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 41s |  The patch does not generate ASF License warnings.  |
|  |   | 353m  2s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.replication.regionserver.TestDrainReplicationQueuesForStandBy |
|   | hadoop.hbase.replication.TestSyncReplicationRemoveRemoteWAL |
|   | hadoop.hbase.regionserver.TestRegionReplicasAreDistributed |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.0 Server=19.03.0 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/631 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux d0e16ebbdec4 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-631/out/precommit/personality/provided.sh |
| git revision | master / bcad0d9f98 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/1/testReport/ |
| Max. process+thread count | 4891 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   3m 24s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :yellow_heart: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 51s |  master passed  |
| :green_heart: |  compile  |   0m 58s |  master passed  |
| :green_heart: |  checkstyle  |   1m 28s |  master passed  |
| :green_heart: |  shadedjars  |   4m 59s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 36s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 25s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 23s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 29s |  the patch passed  |
| :green_heart: |  compile  |   0m 57s |  the patch passed  |
| :green_heart: |  javac  |   0m 57s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 28s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 56s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m  5s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 36s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  | 213m 21s |  hbase-server in the patch passed.  |
| :green_heart: |  asflicense  |   0m 26s |  The patch does not generate ASF License warnings.  |
|  |   | 277m 20s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/631 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 9a19e3107504 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-631/out/precommit/personality/provided.sh |
| git revision | master / 20bfb43db6 |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/2/testReport/ |
| Max. process+thread count | 4683 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by infraio** [comment]
> IIRC there is way to interrupt an existing TRSP from SCP, I think in that method we should also do the same thing?

Set forceNewPlan = true too when call serverCrashed. The TRSP may in any state when call this. 
1. If in REGION_STATE_TRANSITION_GET_ASSIGN_CANDIDATE, forceNewPlan is true then it will choose a round-robin server.
2. If in REGION_STATE_TRANSITION_OPEN, regionClosedAbnormally will set region location to null then it will go back to REGION_STATE_TRANSITION_GET_ASSIGN_CANDIDATE.
3. If in REGION_STATE_TRANSITION_CONFIRM_OPENED, it is same to failed open region.
4. If in REGION_STATE_TRANSITION_CONFIRM_CLOSED, it will go back to REGION_STATE_TRANSITION_GET_ASSIGN_CANDIDATE.
5. If in REGION_STATE_TRANSITION_CLOSE, it will go back to REGION_STATE_TRANSITION_GET_ASSIGN_CANDIDATE.

**Comment by Apache9** [comment]
The problem for setting forceNewPlan = true in serverCrashed is that, the flag will not be persistent so if there is a master crash then the flag will be lost.

But anyway, this is only a nice to have, which means it will not effect correctness so maybe it is fine, but we'd better add some comments about this.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m 17s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :yellow_heart: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   6m 52s |  master passed  |
| :green_heart: |  compile  |   1m 13s |  master passed  |
| :green_heart: |  checkstyle  |   1m 47s |  master passed  |
| :green_heart: |  shadedjars  |   5m 40s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 43s |  master passed  |
| :blue_heart: |  spotbugs  |   5m  2s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 59s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   7m  0s |  the patch passed  |
| :green_heart: |  compile  |   1m 17s |  the patch passed  |
| :green_heart: |  javac  |   1m 17s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 40s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m 50s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  21m 28s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 46s |  the patch passed  |
| :green_heart: |  findbugs  |   5m 28s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 259m  9s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 35s |  The patch does not generate ASF License warnings.  |
|  |   | 333m 52s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestFromClientSideWithCoprocessor |
|   | hadoop.hbase.client.TestAdmin2 |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.7 Server=18.09.7 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/631 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 8a94a1937e35 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-631/out/precommit/personality/provided.sh |
| git revision | master / a0e8723b73 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/3/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/3/testReport/ |
| Max. process+thread count | 4684 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/3/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 39s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :yellow_heart: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 29s |  master passed  |
| :green_heart: |  compile  |   0m 57s |  master passed  |
| :green_heart: |  checkstyle  |   1m 16s |  master passed  |
| :green_heart: |  shadedjars  |   4m 52s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 38s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 38s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 35s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   6m  5s |  the patch passed  |
| :green_heart: |  compile  |   1m 11s |  the patch passed  |
| :green_heart: |  javac  |   1m 11s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 28s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m 39s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m 51s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 38s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 14s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  | 160m 11s |  hbase-server in the patch passed.  |
| :green_heart: |  asflicense  |   0m 32s |  The patch does not generate ASF License warnings.  |
|  |   | 222m 59s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/631 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 05f79c16e557 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-631/out/precommit/personality/provided.sh |
| git revision | master / db0049558e |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/4/testReport/ |
| Max. process+thread count | 4602 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-631/4/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/TransitRegionStateProcedure.java [hunk 1]
```diff
@@ -545,8 +550,15 @@ private static TransitRegionStateProcedure setOwner(MasterProcedureEnv env,
...
   public static TransitRegionStateProcedure assign(MasterProcedureEnv env, RegionInfo region,
       @Nullable ServerName targetServer) {
-    return setOwner(env,
-      new TransitRegionStateProcedure(env, region, targetServer, false, TransitionType.ASSIGN));
+    return assign(env, region, false, targetServer);
+  }
+
+  public static TransitRegionStateProcedure assign(MasterProcedureEnv env, RegionInfo region,
+      boolean forceNewPlan,
+      @Nullable
```
Comment (Apache9) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/TransitRegionStateProcedure.java]: The format is a bit strange?

**Pull request merged**